### ISSUE TITLE
mesmer module naming and structure

### ISSUE THREAD

**Issue by mathause** [MESMER-group/mesmer#348]
I feel like I am being held up by this bike-shedding problem and I would like to settle this before releasing the next version. @znicholls, @jschwaab & @yquilcaille - do you have an opinion?

### 1. module name for climate data processing?

The statistical functions are collected under `mesmer.stats.*`. I like this because its short and descriptive. So far I have failed to come up with a similarly good name for the climate data processing. Currently these functions are under

```python
mesmer.geospatial.*
mesmer.grid.*
mesmer.mask.*
mesmer.weighted.*
```

It feels a bit asymmetric... **Should these be moved to a common module? If yes - how should it be called?** I list some ideas, although none entirely convinces me (**happy to take suggestions**)...

```python
mesmer.cdp.*       # climate data processing -> makes sense, does not read so well
mesmer.cdt.*       # climate data tools      -> makes sense, does not read so well
mesmer.xcm.*       # xr climate model        -> makes only a bit sense, reads well
mesmer.xcmd.*      # xr climate model data   -> makes a bit more sense, reads a bit less well
# while these are more expressive they seem rather long 
mesmer.dataproc.*
mesmer.data_processing.*
mesmer.climate_data_tools.*
mesmer.clim_dta_tools.*
mesmer.process.*
```

Similar existing libraries & programs are cdo, xESMF, xgcm, xmip... 

### 2. where should public functions be exposed?

Currently functions etc. are exposed in their own module, e.g. `mesmer.stats.auto_regression.fit_auto_regression`. This has the advantage that there is a better separation between them. However, for many functions, there is duplication in the name (e.g. `2 x "auto_regression"`) and we need good names for all of them. Alternatively we could expose them as `mesmer.stats.fit_auto_regression`. Thus, the question - **should all functions be moved one level down?**

Instead

```python
mesmer.stats.linear_regression.LinearRegression()
mesmer.stats.smoothing.lowess(...)
mesmer.stats.auto_regression.select_ar_order(...)
mesmer.stats.auto_regression.fit_auto_regression(...)
```

we would go with

```python
mesmer.stats.LinearRegression()
mesmer.stats.lowess(...)
mesmer.stats.select_ar_order(...)
mesmer.stats.fit_auto_regression(...)
```

**Comment by yquilcaille** [comment]
I agree that the module stats would be much bigger than each of these four ones. In those that you suggest, xcmd is not bad. Following our discussion, i suggest as well:
```mesmer.spatial.*       # all of them are related to spatial treatment of data, but still a bit vague
mesmer.geo.*       # another way to say that, although a bit farfetched
mesmer.xgeo.*       # putting some x in there
mesmer.data_handling.*       # following up on your idea about processing, this module could be extended to other functions
mesmer.climproc.*       # coming back to what it does, climate data processing
mesmer.xproc.*       # putting some x in there, sounds weird to me
```


Regarding the 2nd question, I prefer the 2nd solution. Having the main class with this name, and other functions along. Most users would stick to the main classes, some may dig deeper with these other functions, but none of them would be bothered with some duplication in the name. So yes on my side, one level down.

**Comment by znicholls** [comment]
tl;dr - if this is a blocker, pick whatever and move on, it doesn't matter while we have so few users (or at least users giving feedback).

---

I would say it doesn't matter too much. With clear/good versioning, we can always warn users if there are going to be changes so none of these choices really locks us in. Put another way, we shouldn't let this be a blocker. 

Quick thoughts:

- Question 1: if you can't see a good common name/grouping, leave them where they are. There aren't so many that is impossible to navigate
- Question 2: Yes move them all down (current set up is overkill, at least that is my instinct)

I reiterate though: I wouldn't let this be a blocker and I really don't mind what the decision is.

**Comment by jschwaab** [comment]
Regarding question 1: I personally would prefer the ones that make sense or are more expressive. My favorite ones are mesmer.cdt.* and mesmer.climate_data_tools.*

Regarding question 2: I would agree with Yann and Zeb to move them one level down
